[quant][graphmode][fx] Add support for one value being quantized with different qconfigs

jerryzh168 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #54928 [quant][graphmode][fx][refactor] Quantize by Use of a Tensor instead of Tensor
* #54924 [quant][graphmode][fx] Produce torch.cat instead of torch.ops.quantized.cat
* #54813 [quant][graphmode][fx] Optimize cat
* #54812 [nn] Add allow_duplicate option for named_modules
* #54733 [quant][graphmode][fx][refactor] Factor out insert_observers_for_model to a separate function
* #54644 [quant][graphmode][fx] Separate handling Copy operator to a helper function
* #54643 [quant][fx][graphmode][refactor] Change activation_post_process_map to track the observer name instead
* #54073 [quant][graphmode][refactor] Remove reduandent code
* **#53586 [quant][graphmode][fx] Add support for one value being quantized with different qconfigs**

Summary:
Previously one value can only be quantized to one dtype, this PR adds the support for quantizing one value
in the fx graph with multiple dtypes, e.g. first quantize to int8 and then float16

might do some followup PRs to clean up the hacks and refactor the code.

Test Plan:
python test/test_quantization.py TestQuantizeFx.test_multiple_qconfigs_single_value

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D26912676](https://our.internmc.facebook.com/intern/diff/D26912676)

jerryzh168 committed:
[quant][graphmode][fx] Add support for one value being quantized with different qconfigs

Summary:
Previously one value can only be quantized to one dtype, this PR adds the support for quantizing one value
in the fx graph with multiple dtypes, e.g. first quantize to int8 and then float16

might do some followup PRs to clean up the hacks and refactor the code.

Test Plan:
python test/test_quantization.py TestQuantizeFx.test_multiple_qconfigs_single_value

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][fx] Add support for one value being quantized with different qconfigs"


Summary:
Previously one value can only be quantized to one dtype, this PR adds the support for quantizing one value
in the fx graph with multiple dtypes, e.g. first quantize to int8 and then float16

might do some followup PRs to clean up the hacks and refactor the code.

Test Plan:
python test/test_quantization.py TestQuantizeFx.test_multiple_qconfigs_single_value

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

vkuzo commented on torch/quantization/fx/quantize.py:
```diff
@@ -78,7 +80,7 @@
         node: Node, observer: torch.quantization.ObserverBase,
         model: torch.nn.Module,
-        activation_post_process_map: Dict[str, torch.quantization.ObserverBase],
+        activation_post_process_map: Dict[str, List[torch.quantization.ObserverBase]],
```
does the graph look like

```
          --> obs2
         /
node --> obs1
```

or 

```
node -> obs1 -> obs2
```

?

jerryzh168 replied:
the second one

jerryzh168 commented on torch/quantization/fx/quantization_patterns.py:
```diff
@@ -181,7 +182,7 @@ def convert(self, quantizer: QuantizerCls, node: Node, load_arg: Callable,
                     load_arg(quantized=[quantized_index])(self.binary_op_node.args), self.binary_op_node.kwargs)
             else:
-                activation_post_process = quantizer.activation_post_process_map[node.name]
+                activation_post_process = quantizer.activation_post_process_map[node.name].pop(0)
```
yes, this is a bit hacky, I plan to clean up in a later PR

vkuzo replied:
makes sense, any chance we could make it happen in this PR, or move the refactor to be before this PR?  This seems like a pretty risky change

jerryzh168 replied:
I can't think of a easy way currently without making larger changes. I think it should be fine as long as all the tests passes. but please let me know if you have any suggestions.

vkuzo replied:
hmm, can we just do something like

```
cur_obs_idx = 0
... when needed, increment cur_obs_idx
cur_obs = quantizer.activation_post_process_map[node.name][cur_obs_idx]
```

it's a bit of extra bookkeeping, but it removes state mutations from this variable and instead moves them into the code

vkuzo commented on torch/quantization/fx/quantize.py:
```diff
@@ -1096,10 +1112,10 @@ def _find_quants(self, graph: Graph, matches: Dict[str, MatchResult],
 
         Outputs a map of
-         node_name -> (QuantizeHandler instance (always DefaultQuantizeHandler),
+         node_name -> list of (QuantizeHandler instance (always DefaultQuantizeHandler),
```
can we make sure that the docblock explains why this is a list

jerryzh168 replied:
sure

vkuzo commented on torch/quantization/fx/quantization_patterns.py:
```diff
@@ -927,7 +929,12 @@ class CopyNode(QuantizeHandler):
-        return quantizer.quantized_graph.node_copy(node, load_arg(quantized=None))
+        qconfig = quantizer.qconfig_map[node.name]
+        if activation_is_int8_quantized(qconfig):
+            return quantizer.quantized_graph.node_copy(node, load_arg(quantized=None))
+        else:
+            # fp16 static quantization
+            return quantizer.quantized_graph.node_copy(node, load_arg(quantized=False))
```
wondering on what's a good way to explain this?  Does `quantized` from `load_arg` mean the same thing as `activation_is_int8_quantized`?  If not, should we rename the `load_arg` argument to be more clear?

jerryzh168 replied:
yeah this is a bit confusing, I'll think about how to make this more clear

vkuzo commented on torch/quantization/fx/quantize.py:
```diff
@@ -669,12 +674,13 @@ def _convert(self, model: GraphModule, is_reference: bool = False,
         env: Dict[str, Node] = {}
-        quant_env: Dict[str, Node] = {}
+        # TODO: merge quant_env with env
+        quant_env: Dict[str, Tuple[Node, torch.dtype]] = {}
```
can we keep the types of `env` and `quant_env` consistent?  If `quant_env` needs a dtype, can we add the `torch.float` dtype to every entry of `env?

jerryzh168 replied:
yeah, see TODO comment above, I plan to merge them later

jerryzh168 committed:
Update on "[quant][graphmode][fx] Add support for one value being quantized with different qconfigs"


Summary:
Previously one value can only be quantized to one dtype, this PR adds the support for quantizing one value
in the fx graph with multiple dtypes, e.g. first quantize to int8 and then float16

might do some followup PRs to clean up the hacks and refactor the code.

Test Plan:
python test/test_quantization.py TestQuantizeFx.test_multiple_qconfigs_single_value

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D26912676](https://our.internmc.facebook.com/intern/diff/D26912676)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][fx] Add support for one value being quantized with different qconfigs"


Summary:
Previously one value can only be quantized to one dtype, this PR adds the support for quantizing one value
in the fx graph with multiple dtypes, e.g. first quantize to int8 and then float16

might do some followup PRs to clean up the hacks and refactor the code.

Test Plan:
python test/test_quantization.py TestQuantizeFx.test_multiple_qconfigs_single_value

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D26912676](https://our.internmc.facebook.com/intern/diff/D26912676)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][fx] Add support for one value being quantized with different qconfigs"


Summary:
Previously one value can only be quantized to one dtype, this PR adds the support for quantizing one value
in the fx graph with multiple dtypes, e.g. first quantize to int8 and then float16

might do some followup PRs to clean up the hacks and refactor the code.

Test Plan:
python test/test_quantization.py TestQuantizeFx.test_multiple_qconfigs_single_value

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D26912676](https://our.internmc.facebook.com/intern/diff/D26912676)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][fx] Add support for one value being quantized with different qconfigs"


Summary:
Previously one value can only be quantized to one dtype, this PR adds the support for quantizing one value
in the fx graph with multiple dtypes, e.g. first quantize to int8 and then float16

might do some followup PRs to clean up the hacks and refactor the code.

Test Plan:
python test/test_quantization.py TestQuantizeFx.test_multiple_qconfigs_single_value

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D26912676](https://our.internmc.facebook.com/intern/diff/D26912676)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][fx] Add support for one value being quantized with different qconfigs"


Summary:
Previously one value can only be quantized to one dtype, this PR adds the support for quantizing one value
in the fx graph with multiple dtypes, e.g. first quantize to int8 and then float16

might do some followup PRs to clean up the hacks and refactor the code.

Test Plan:
python test/test_quantization.py TestQuantizeFx.test_multiple_qconfigs_single_value

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D26912676](https://our.internmc.facebook.com/intern/diff/D26912676)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][fx] Add support for one value being quantized with different qconfigs"


Summary:
Previously one value can only be quantized to one dtype, this PR adds the support for quantizing one value
in the fx graph with multiple dtypes, e.g. first quantize to int8 and then float16

might do some followup PRs to clean up the hacks and refactor the code.

Test Plan:
python test/test_quantization.py TestQuantizeFx.test_multiple_qconfigs_single_value

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D26912676](https://our.internmc.facebook.com/intern/diff/D26912676)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][fx] Add support for one value being quantized with different qconfigs"


Summary:
Previously one value can only be quantized to one dtype, this PR adds the support for quantizing one value
in the fx graph with multiple dtypes, e.g. first quantize to int8 and then float16

might do some followup PRs to clean up the hacks and refactor the code.

Test Plan:
python test/test_quantization.py TestQuantizeFx.test_multiple_qconfigs_single_value

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D26912676](https://our.internmc.facebook.com/intern/diff/D26912676)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][fx] Add support for one value being quantized with different qconfigs"


Summary:
Previously one value can only be quantized to one dtype, this PR adds the support for quantizing one value
in the fx graph with multiple dtypes, e.g. first quantize to int8 and then float16

might do some followup PRs to clean up the hacks and refactor the code.

Test Plan:
python test/test_quantization.py TestQuantizeFx.test_multiple_qconfigs_single_value

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D26912676](https://our.internmc.facebook.com/intern/diff/D26912676)

[ghstack-poisoned]

vkuzo commented on test/quantization/test_quantize_fx.py:
```diff
@@ -634,6 +634,7 @@ def forward(self, x):
         self.assertEqual(model_device, device)
 
     @skipIfNoFBGEMM
+    @unittest.skip("dict doesnot work in fx?")
```
is it related to this PR?

jerryzh168 replied:
this test fails after the change, will need to ask FX team to help debug the issue

vkuzo commented on torch/quantization/fx/quantize.py:
```diff
@@ -81,15 +83,19 @@
          node: Node
          observer: observer/fake_quantize module instance
     """
+    if not model.training and isinstance(observer, torch.quantization.FixedQParamsFakeQuantize):
+        return
```
just curious, what's the context on this one?  Maybe we can add a comment?

jerryzh168 replied:
Sure I can add a comment. reason is fixedqparams fake quant only needs to be there for training, we can either do this check here or in _find_quant.

vkuzo reviewed (approved):
makes sense, accepting to unblock, would be great to undo the additional state in the future.  One idea I had when reading this PR...thoughts about this?
```
# map from (node.name, qconfig) to observer
self.activation_post_process_map: Dict[Tuple(str, QConfig), ObserverBase]
```
If the only use case for multiple observers for a node is multiple qconfigs, perhaps that can be used to deduplicate, and we wouldn't need to add an additional indices variable?  I haven't thought through it deeply, more asking for your early opinion.

jerryzh168 committed:
Update on "[quant][graphmode][fx] Add support for one value being quantized with different qconfigs"


Summary:
Previously one value can only be quantized to one dtype, this PR adds the support for quantizing one value
in the fx graph with multiple dtypes, e.g. first quantize to int8 and then float16

might do some followup PRs to clean up the hacks and refactor the code.

Test Plan:
python test/test_quantization.py TestQuantizeFx.test_multiple_qconfigs_single_value

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D26912676](https://our.internmc.facebook.com/intern/diff/D26912676)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][fx] Add support for one value being quantized with different qconfigs"


Summary:
Previously one value can only be quantized to one dtype, this PR adds the support for quantizing one value
in the fx graph with multiple dtypes, e.g. first quantize to int8 and then float16

might do some followup PRs to clean up the hacks and refactor the code.

Test Plan:
python test/test_quantization.py TestQuantizeFx.test_multiple_qconfigs_single_value

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D26912676](https://our.internmc.facebook.com/intern/diff/D26912676)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][fx] Add support for one value being quantized with different qconfigs"


Summary:
Previously one value can only be quantized to one dtype, this PR adds the support for quantizing one value
in the fx graph with multiple dtypes, e.g. first quantize to int8 and then float16

might do some followup PRs to clean up the hacks and refactor the code.

Test Plan:
python test/test_quantization.py TestQuantizeFx.test_multiple_qconfigs_single_value

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D26912676](https://our.internmc.facebook.com/intern/diff/D26912676)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][fx] Add support for one value being quantized with different qconfigs"


Summary:
Previously one value can only be quantized to one dtype, this PR adds the support for quantizing one value
in the fx graph with multiple dtypes, e.g. first quantize to int8 and then float16

might do some followup PRs to clean up the hacks and refactor the code.

Test Plan:
python test/test_quantization.py TestQuantizeFx.test_multiple_qconfigs_single_value

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D26912676](https://our.internmc.facebook.com/intern/diff/D26912676)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][fx] Add support for one value being quantized with different qconfigs"


Summary:
Previously one value can only be quantized to one dtype, this PR adds the support for quantizing one value
in the fx graph with multiple dtypes, e.g. first quantize to int8 and then float16

might do some followup PRs to clean up the hacks and refactor the code.

Test Plan:
python test/test_quantization.py TestQuantizeFx.test_multiple_qconfigs_single_value

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D26912676](https://our.internmc.facebook.com/intern/diff/D26912676)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][fx] Add support for one value being quantized with different qconfigs"


Summary:
Previously one value can only be quantized to one dtype, this PR adds the support for quantizing one value
in the fx graph with multiple dtypes, e.g. first quantize to int8 and then float16

might do some followup PRs to clean up the hacks and refactor the code.

Test Plan:
python test/test_quantization.py TestQuantizeFx.test_multiple_qconfigs_single_value

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D26912676](https://our.internmc.facebook.com/intern/diff/D26912676)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][fx] Add support for one value being quantized with different qconfigs"


Summary:
Previously one value can only be quantized to one dtype, this PR adds the support for quantizing one value
in the fx graph with multiple dtypes, e.g. first quantize to int8 and then float16

might do some followup PRs to clean up the hacks and refactor the code.

Test Plan:
python test/test_quantization.py TestQuantizeFx.test_multiple_qconfigs_single_value

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D26912676](https://our.internmc.facebook.com/intern/diff/D26912676)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][fx] Add support for one value being quantized with different qconfigs"


Summary:
Previously one value can only be quantized to one dtype, this PR adds the support for quantizing one value
in the fx graph with multiple dtypes, e.g. first quantize to int8 and then float16

might do some followup PRs to clean up the hacks and refactor the code.

Test Plan:
python test/test_quantization.py TestQuantizeFx.test_multiple_qconfigs_single_value

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D26912676](https://our.internmc.facebook.com/intern/diff/D26912676)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][fx] Add support for one value being quantized with different qconfigs"


Summary:
Previously one value can only be quantized to one dtype, this PR adds the support for quantizing one value
in the fx graph with multiple dtypes, e.g. first quantize to int8 and then float16

might do some followup PRs to clean up the hacks and refactor the code.

Test Plan:
python test/test_quantization.py TestQuantizeFx.test_multiple_qconfigs_single_value

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D26912676](https://our.internmc.facebook.com/intern/diff/D26912676)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][fx] Add support for one value being quantized with different qconfigs"


Summary:
Previously one value can only be quantized to one dtype, this PR adds the support for quantizing one value
in the fx graph with multiple dtypes, e.g. first quantize to int8 and then float16

might do some followup PRs to clean up the hacks and refactor the code.

Test Plan:
python test/test_quantization.py TestQuantizeFx.test_multiple_qconfigs_single_value

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D26912676](https://our.internmc.facebook.com/intern/diff/D26912676)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][fx] Add support for one value being quantized with different qconfigs"


Summary:
Previously one value can only be quantized to one dtype, this PR adds the support for quantizing one value
in the fx graph with multiple dtypes, e.g. first quantize to int8 and then float16

might do some followup PRs to clean up the hacks and refactor the code.

Test Plan:
python test/test_quantization.py TestQuantizeFx.test_multiple_qconfigs_single_value

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D26912676](https://our.internmc.facebook.com/intern/diff/D26912676)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][fx] Add support for one value being quantized with different qconfigs"


Summary:
Previously one value can only be quantized to one dtype, this PR adds the support for quantizing one value
in the fx graph with multiple dtypes, e.g. first quantize to int8 and then float16

might do some followup PRs to clean up the hacks and refactor the code.

Test Plan:
python test/test_quantization.py TestQuantizeFx.test_multiple_qconfigs_single_value

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D26912676](https://our.internmc.facebook.com/intern/diff/D26912676)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][fx] Add support for one value being quantized with different qconfigs"


Summary:
Previously one value can only be quantized to one dtype, this PR adds the support for quantizing one value
in the fx graph with multiple dtypes, e.g. first quantize to int8 and then float16

might do some followup PRs to clean up the hacks and refactor the code.

Test Plan:
python test/test_quantization.py TestQuantizeFx.test_multiple_qconfigs_single_value

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D26912676](https://our.internmc.facebook.com/intern/diff/D26912676)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][fx] Add support for one value being quantized with different qconfigs"


Summary:
Previously one value can only be quantized to one dtype, this PR adds the support for quantizing one value
in the fx graph with multiple dtypes, e.g. first quantize to int8 and then float16

might do some followup PRs to clean up the hacks and refactor the code.

Test Plan:
python test/test_quantization.py TestQuantizeFx.test_multiple_qconfigs_single_value

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D26912676](https://our.internmc.facebook.com/intern/diff/D26912676)

[ghstack-poisoned]
